actors/builtin/miner/deadline_state.go

```diff
@@ -502,3 +504,161 @@ func (dl *Deadline) TerminateSectors(
+		if _, ok := toRemoveSet[uint64(partIdx)]; !ok {
+			return newPartitions.AppendContinuous(&lazyPartition)
+		}
+
+		// Ok, actually unmarshal the partition.
+		byteReader.Reset(lazyPartition.Raw)
+		err := partition.UnmarshalCBOR(&byteReader)
+		byteReader.Reset(nil)
+		if err != nil {
+			return xerrors.Errorf("failed to decode partition %d: %w", partIdx, err)
+		}
+
+		// Don't allow removing partitions with faulty sectors.
+		hasNoFaults, err := partition.Faults.IsEmpty()
+		if err != nil {
+			return xerrors.Errorf("failed to decode faults for partition %d: %w", partIdx, err)
+		}
+		if !hasNoFaults {
+			// TODO: this is an invalid argument error.
+			// https://github.com/filecoin-project/specs-actors/issues/597
+			return xerrors.Errorf("cannot remove partition %d: has faults", partIdx)
+		}
+
+		// Get the live sectors.
+		liveSectors, err := partition.LiveSectors()
+		if err != nil {
+			return xerrors.Errorf("failed to calculate live sectors for partition %d: %w", partIdx, err)
+		}
+
+		allDeadSectors = append(allDeadSectors, partition.Terminated)
+		allLiveSectors = append(allLiveSectors, liveSectors)
+		removedPower = removedPower.Add(partition.LivePower)
+		return nil
+	}); err != nil {
+		return nil, nil, NewPowerPairZero(), xerrors.Errorf("while removing partitions: %w", err)
+	}
+
+	dl.Partitions, err = newPartitions.Root()
+	if err != nil {
+		return nil, nil, NewPowerPairZero(), xerrors.Errorf("failed to persist new partition table: %w", err)
+	}
+
+	dead, err = bitfield.MultiMerge(allDeadSectors...)
+	if err != nil {
+		return nil, nil, NewPowerPairZero(), xerrors.Errorf("failed to merge dead sector bitfields: %w", err)
+	}
+	live, err = bitfield.MultiMerge(allLiveSectors...)
+	if err != nil {
+		return nil, nil, NewPowerPairZero(), xerrors.Errorf("failed to merge live sector bitfields: %w", err)
+	}
+
+	// Update sector counts.
+	removedDeadSectors, err := dead.Count()
+	if err != nil {
+		return nil, nil, NewPowerPairZero(), xerrors.Errorf("failed to count dead sectors: %w", err)
+	}
+
+	removedLiveSectors, err := live.Count()
+	if err != nil {
+		return nil, nil, NewPowerPairZero(), xerrors.Errorf("failed to count live sectors: %w", err)
+	}
+
+	dl.LiveSectors -= removedLiveSectors
+	dl.TotalSectors -= removedLiveSectors + removedDeadSectors
+
+	// Update expiration bitfields.
+	{
+		expirationEpochs, err := LoadBitfieldQueue(store, dl.ExpirationsEpochs, quant)
+		if err != nil {
+			return nil, nil, NewPowerPairZero(), xerrors.Errorf("failed to load expiration queue: %w", err)
+		}
+
+		var epochsToRemove []uint64
+		err = expirationEpochs.ForEach(func(epoch abi.ChainEpoch, bf *bitfield.BitField) error {
+			bf, err := bitfield.CutBitField(bf, toRemove)
+			if err != nil {
+				return err
+			}
+			if empty, err := bf.IsEmpty(); err != nil {
+				return err
+			} else if !empty {
+				return expirationEpochs.Set(uint64(epoch), bf)
+			}
+			epochsToRemove = append(epochsToRemove, uint64(epoch))
+			return nil
+		})
```

Comment:
Nit: push this down to a method on BitfieldQueue

Reply:
Done.